### DISCUSSION TITLE
tutorials/client-side-graphql-react/06-writing-queries

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#109] [course feedback]
# tutorials/client-side-graphql-react/06-writing-queries

https://www.apollographql.com/tutorials/client-side-graphql-react/06-writing-queries

**Comment by EleazarVillarreal** [comment]
The numberOfViews field is missing from the GET_TRACKS query which will not update the track-card in the mutation exercise.

**Reply by lizhennessy** [reply]
Hi there @EleazarVillarreal! You're correct that this `GET_TRACKS` query does not contain the `numberOfViews` field: in this case, we're concerned just with the data that we're displaying for our tracks on the home page, which does not include the number of views each track has. We'll include this field in the query that we make for each Track card, in [an upcoming lesson](https://www.apollographql.com/tutorials/client-side-graphql-react/08-use-query-with-variables).

**Comment by nishantchapa** [comment]
I am seeing error for track property in tracks.tsx file

Type '{ __typename?: "Track" | undefined; id: string; title: string; thumbnail?: string | null | undefined; length?: number | null | undefined; modulesCount?: number | null | undefined; author: { ...; }; }' is not assignable to type 'Omit<Track, "modules">'.
  Types of property 'author' are incompatible.
    Property 'id' is missing in type '{ __typename?: "Author" | undefined; name: string; photo?: string | null | undefined; }' but required in type 'Author'.ts(2322)
graphql.ts(22, 3): 'id' is declared here.
track-card.tsx(12, 29): The expected type comes from property 'track' which is declared here on type 'IntrinsicAttributes & { track: Omit<Track, "modules">; }'
(property) track: Omit<Track, "modules">

**Reply by lizhennessy** [reply]
Hi there @nishantchapa ! At what step in the lesson does this error appear? And do you have your code pushed somewhere that I can take a look? I wonder if there might be a hiccup in the codegen process.

**Reply by nishantchapa** [reply]
@lizhennessy I noticed that the query we used in explorer (even in the screenshot of the tutorial) doesn't contain id for author, but the query used for GET_TRACKS in gql function has id and hence the confusion where I copied the query from explorer and used in GET_TRACKS

**Comment by kristiyan-velkov** [comment]
Hello,

I noticed that the example code provided lacks the id property, which is used in the TrackCard component. Could you please update the example to include the id?

Here is the original example:

typescript
Copy code
const TrackCard: React.FC<{ track: Omit<Track, "modules"> }> = ({ track }) => {
  const { title, thumbnail, author, length, modulesCount } = track;
  // ... TrackCard body
}

// <CardContainer to={`/track/${id}`}>
Thank you!

This version of your message is clearer and more specific, which should help in getting the example updated appropriately.

**Reply by mabuyo** [reply]
Good catch, thanks @kristiyan-velkov, we will update this!

**Comment by jessie-pann** [comment]
hey, the gql function imported from "../__generated__/" return type unknow: function gql(source: string): unknown, which give the type of GET_TRACKS as unknown, then when passing GET_TRACKS as an argument for useQuery, I get the following error: Argument of type 'unknown' is not assignable to parameter of type 'DocumentNode | TypedDocumentNode<any, OperationVariables>'. 
could anyone help pls? thank you

**Reply by mabuyo** [reply]
Hi @jessie-pann! Hmm so when you hover over your import line, you're getting the error that it can't find the `gql` function you're trying to import? Could you try reviewing the steps in the [previous lesson](https://www.apollographql.com/tutorials/client-side-graphql-react/05-codegen) for codegen and make sure you followed all of them?

**Reply by devi-rbm** [reply]
@jessie-pann, were you able to get past your issue?

**Comment by mrrevil** [comment]
Hi @mabuyo, is it an anti-pattern to use an auto-generated type for a query (e.g GetTracksQuery) as a type for a component prop? I have quite big queries in mind with nested objects where we fetch just some fields from those objects. 

Taking this line from the tutorial
`const TrackCard: React.FC<{ track: Omit<Track, "modules"> }> = ({ track }) => {`

Would you build a dedicated TS type omitting/picking different props from different object levels or would it be a good approach to just reuse a part of the query type which reflects the actual object shape we get from a query so e.g:

`{track: GetTracksQuery['tracksForHome']}`

**Reply by mabuyo** [reply]
Hey there. @mrrevil! I checked with our engineering team and I'm passing along the response, including  a couple resources in our docs. Hope it helps!

> I'd say it's perfectly fine if those props really are what your component needs, and not any additional props that might just be part of the same query. For that purpose, I would recommend fragment colocation:
https://www.apollographql.com/docs/react/data/fragments#colocating-fragments

You can also take a look at using `TypedDocumentNode`: https://www.apollographql.com/docs/react/development-testing/static-typing#using-typeddocumentnode

**Comment by BYAsurim** [comment]
Hey there.  const GET_TRACKS = gql(`
  query GetTracks {
    tracksForHome {
      id
      title
      thumbnail
      length
      modulesCount
      author {
        id
        name
        photo
      }
    }
  }
`);
does this gql function have syntax highlighting for easier operation or should I use explorer to create a query and copy it into the function?

**Reply by lizhennessy** [reply]
Hi @BYAsurim ! If you're using VSCode, there are several extensions that can help you out here. We recommend the [Apollo GraphQL extension](https://marketplace.visualstudio.com/items?itemName=apollographql.vscode-apollo) (syntax highlighting, autocomplete, etc.) as well as the [GraphQL Foundation extension](https://marketplace.visualstudio.com/items?itemName=GraphQL.vscode-graphql).
